### PR TITLE
[Cult 3.0] Conversions & Stuns, Runes V & VI

### PR DESCRIPTION
**Pull request by DeityLink**
<h2>Conversion</h2>

successful: http://is3.4chan.org/vg/1531968761912.webm
imageimageimage
aborted: http://is3.4chan.org/vg/1531989877422.webm
image
failure: http://is3.4chan.org/vg/1531989218888.webm
imageimage

conversions now occur in two phases:
=in both phases=
* If the victim is pulled away from the rune, the ritual immediately ends.
* As long as the ritual proceeds normally, the victim will automatically remain stun, knockdown, & mute.

=Phase 1=
* Summoner must stand adjacent to the rune, if he's away for a total of 3 seconds, the ritual ends
* This phase takes 10 seconds by default. This time will be greatly increased by:
-heads of staff & security taking longer
-implanted people taking even longer (stacks)
-captain & chaplain taking yet even longer (chaplain ritual never succeeds though, always sacrifices him)
* This time will be reduced by wearing cult attire, each piece of equipment having their own ratio

=Phase 2=
* Summoner no longer has to remain adjacent
* Always takes 10 seconds
* If the victim has cult set to Yes or Always in their preferences, they will always become a cultist at the end
* If the victim is cultbanned or has their preferences set to never, they will always get sac'd
* If the victim has cult set to No, a prompt appears at the beginning of the phase offering him either choices. If they haven't made their choice by the end of the phase, they will get sac'd.

If the victim is successfully converted, they will be immediately purged of the stun/knockdown/mute. If the ritual is aborted, the debuffs will linger a bit (not long).

Upon getting sacrificed, the victim leaves behind a box with their stuff, a skull, and a cup filled with some of their blood. (incidentally, added both the coffer and cup as new items)

<h2>Stun</h2>

http://i.4cdn.org/vg/1532034549610.webm
rune:
image
talisman aoe:
image
talisman touch:
image

clean version of the stun talisman (with sound: https://www.dropbox.com/s/69fg515kzx4umqn/dreamseeker%202018-07-20%2003-30-45-27.webm?dl=0 )

Stun runes and their new talisman aoe version now stun a second (10 ticks) after getting activated. Beside that the stun works just like before except:
* it stuns longer (rune 6 sec, talisman 4 sec, cultists get up 1 sec faster)
* the talisman aoe version only has a 4 tiles range.

The stun talismans work just like before except:
* their incantation is now whispered
* only 5% chance to use the Touhou reference
* that animation (which will be either toned down or hidden if you have the Silent Casting tattoo)

both the conversion and stun sounds can only be heard at a very short range.

### CODE REVIEW

**Comment by DeityLink** [comment]
The delay means that by running fast you can trigger and barely avoid the rune's stun radius. Even more easily when using a talisman.

**Comment by DeityLink** [comment]
Added a fullscreen "flash" for victims during conversion. The red becomes more opaque as the conversion progresses. 
image

Slimes getting sacrificed leave behind a cup of slime jelly. Aliens leave behind a cup of radium.

File: code/datums/gamemode/factions/bloodcult/bloodcult_items.dm [hunk 1]
```diff
@@ -422,29 +451,55 @@ var/list/arcane_tomes = list()
...
+
+/obj/item/weapon/storage/backpack/cultpack/attack_self(mob/user as mob)
+	..()
+	if(skulls)
+		for(,skulls > 0,skulls--)
+			new/obj/item/weapon/skull(get_turf(src))
+		update_icon(user)
+
+/obj/item/weapon/storage/backpack/cultpack/attackby(obj/item/weapon/W as obj, mob/user as mob)
+	if(W == src)
```
Comment (DamianX) [inline code/datums/gamemode/factions/bloodcult/bloodcult_items.dm]: how can this happen?

File: code/datums/gamemode/factions/bloodcult/bloodcult_items.dm [hunk 2]
```diff
@@ -273,16 +273,26 @@ var/list/arcane_tomes = list()
...
 	else
 		to_chat(user, "<span class='info'>This one was some arcane drawings on it. You cannot read them.</span>")
 
 /obj/item/weapon/talisman/attack_self(var/mob/living/user)
 	if (iscultist(user))
 		trigger(user)
 
+/obj/item/weapon/talisman/attack(var/mob/living/target, var/mob/living/user)
+	if(iscultist(user))
+		if(spell_type)
```
Comment (MadmanMartian) [inline code/datums/gamemode/factions/bloodcult/bloodcult_items.dm]: Why not combine the two if statements if neither have an else?

File: code/datums/gamemode/factions/bloodcult/bloodcult_items.dm [hunk 3]
```diff
@@ -422,29 +451,55 @@ var/list/arcane_tomes = list()
...
+	name = "trophy rack"
+	desc = "It's useful for both carrying extra gear and proudly declaring your insanity."
+	icon_state = "cultpack_0skull"
+	item_state = "cultpack_0skull"
+	var/skulls = 0
+
+/obj/item/weapon/storage/backpack/cultpack/attack_self(mob/user as mob)
+	..()
+	if(skulls)
+		for(,skulls > 0,skulls--)
```
Comment (MadmanMartian) [inline code/datums/gamemode/factions/bloodcult/bloodcult_items.dm]: for(var/i = 0 to skulls) perhaps?

File: code/datums/gamemode/factions/bloodcult/bloodcult_items.dm [hunk 4]
```diff
@@ -422,29 +451,55 @@ var/list/arcane_tomes = list()
...
+			new/obj/item/weapon/skull(get_turf(src))
+		update_icon(user)
+
+/obj/item/weapon/storage/backpack/cultpack/attackby(obj/item/weapon/W as obj, mob/user as mob)
```
Comment (MadmanMartian) [inline code/datums/gamemode/factions/bloodcult/bloodcult_items.dm]: as (thing) isn't necessary.

File: code/datums/gamemode/factions/bloodcult/bloodcult_runespells.dm [hunk 5]
```diff
@@ -537,23 +590,437 @@
...
+				playsound(R, 'sound/effects/convert_process.ogg', 10, 0, -4)
+				//then progress through the ritual
+				victim.Silent(5)
+				victim.Knockdown(5)
+				victim.Stun(5)
+				var/progress = 10//10 seconds to reach second phase for a naked cultist
+				progress += activator.get_cult_power()//down to 1-2 seconds when wearing cult gear
+				var/delay = 0
+				for(var/obj/item/weapon/implant/I in victim)
+					if(I.implanted && istype(I,/obj/item/weapon/implant/loyalty))
```
Comment (DamianX) [inline code/datums/gamemode/factions/bloodcult/bloodcult_runespells.dm]: why not just look for `/obj/item/weapon/implant/loyalty` in the for loop?

File: code/datums/gamemode/factions/bloodcult/bloodcult_items.dm [hunk 6]
```diff
@@ -483,12 +589,45 @@ var/list/arcane_tomes = list()
...
+	icon_state = "cult"
+	item_state = "syringe_kit"
+	starting_materials = list(MAT_IRON = 3750)
+	w_type=RECYK_METAL
+
+///////////////////////////////////////CULT GLASS////////////////////////////////////////////////
+
+/obj/item/weapon/reagent_containers/food/drinks/cult
+	name = "cup"
+	desc = "An spooky looking cup with a skull motif."
```
Comment (MadmanMartian) [inline code/datums/gamemode/factions/bloodcult/bloodcult_items.dm]: A spooky looking cup with a skull motif.

File: code/datums/gamemode/factions/bloodcult/bloodcult_runespells.dm [hunk 7]
```diff
@@ -67,14 +72,27 @@
 /datum/rune_spell/proc/midcast(var/mob/add_cultist)
 	return
 
+/datum/rune_spell/proc/blood_pay()
+	var/data = use_available_blood(activator, cost_invoke)
+	if (data["result"] == "failure")
```
Comment (MadmanMartian) [inline code/datums/gamemode/factions/bloodcult/bloodcult_runespells.dm]: I can see this going weird in the future. Put "result" and "failure" and the like from use_available_blood as defines.

File: code/datums/gamemode/factions/bloodcult/bloodcult_runespells.dm [hunk 8]
```diff
@@ -67,14 +72,27 @@
...
+	return
+
 /datum/rune_spell/proc/cast()
 	spell_holder.visible_message("<span class='warning'>This rune wasn't properly set up, tell a coder.</span>")
 	qdel(src)
 
-/datum/rune_spell/proc/abort(var/cause = "erased")
+/datum/rune_spell/proc/abort(var/cause)
 	switch (cause)
 		if ("erased")
```
Comment (MadmanMartian) [inline code/datums/gamemode/factions/bloodcult/bloodcult_runespells.dm]: Same with these. Defines!

File: code/modules/mob/living/living.dm [hunk 9]
```diff
@@ -1733,3 +1733,9 @@ Thanks.
 		"suiciding")
 
 	reset_vars_after_duration(resettable_vars, duration)
+
+/mob/living/proc/Silent(amount)
+	silent = max(max(silent,amount),0)
+
+/mob/living/proc/SetSilent(amount)
+	knockdown = max(amount,0)
```
Comment (DamianX) [inline code/modules/mob/living/living.dm]: >knockdown

File: code/datums/gamemode/factions/bloodcult/bloodcult_runespells.dm [hunk 10]
```diff
@@ -537,23 +590,437 @@
...
+	victim.Stun(5)
+	victim.overlay_fullscreen("conversionborder", /obj/abstract/screen/fullscreen/conversion_border)
+	victim.overlay_fullscreen("conversionred", /obj/abstract/screen/fullscreen/conversion_red)
+	victim.update_fullscreen_alpha("conversionred", 255, 5)
+	victim.update_fullscreen_alpha("conversionborder", 255, 5)
+	conversion = new(T)
+	flick("rune_convert_start",conversion)
+	playsound(R, 'sound/effects/convert_start.ogg', 75, 0, -4)
+
+	if (victim.mind)
```
Comment (MadmanMartian) [inline code/datums/gamemode/factions/bloodcult/bloodcult_runespells.dm:676]: You don't seem to have actually increased the remaining time in the event the person is in impede_medium, impede_hard, or implanted, despite this reading as it being the case.

File: code/datums/gamemode/factions/bloodcult/bloodcult_runespells.dm [hunk 11]
```diff
@@ -537,23 +590,437 @@
...
+	for(var/obj/item/weapon/implant/I in victim)
+		if(I.implanted && istype(I,/obj/item/weapon/implant/loyalty))
+			to_chat(victim, "<span class='warning'>Your loyalty implants drastically slows down the ritual's progression.</span>")
+			to_chat(activator, "<span class='warning'>Their mind seems to reject the ritual by reflex. The ritual will take much longer.</span>")
+			break
+
+
+	spawn()
+		while (remaining > 0)
+			if (destroying_self || !spell_holder || !activator || !victim)
```
Comment (DamianX) [inline code/datums/gamemode/factions/bloodcult/bloodcult_runespells.dm:694]: `destroying_self` was not added in this PR but did you not know about `gcDestroyed`?

File: code/datums/gamemode/factions/bloodcult/bloodcult_runespells.dm [hunk 12]
```diff
@@ -537,23 +590,437 @@
...
+		if (victim.mind.assigned_role in impede_medium)
+			to_chat(victim, "<span class='warning'>Your sense of duty impedes down the ritual.</span>")
+			to_chat(activator, "<span class='warning'>Their will is strong, the ritual will take longer.</span>")
+
+		if (victim.mind.assigned_role in impede_hard)
+			to_chat(victim, "<span class='warning'>Your devotion to Nanotrasen impedes the ritual.</span>")
+			to_chat(activator, "<span class='warning'>Their willpower is amazing, the ritual will be exhausting.</span>")
+
+	for(var/obj/item/weapon/implant/I in victim)
+		if(I.implanted && istype(I,/obj/item/weapon/implant/loyalty))
```
Comment (DamianX) [inline code/datums/gamemode/factions/bloodcult/bloodcult_runespells.dm]: why not just look for `/obj/item/weapon/implant/loyalty` in the for loop?

File: code/datums/gamemode/role/cultist.dm [hunk 13]
```diff
@@ -45,13 +45,17 @@
 			to_chat(antag.current, "<span class='sinister'>Assist your new compatriots in their dark dealings. Their goal is yours, and yours is theirs. You serve the Dark One above all else. Bring It back.</span>")
 		if ("custom")
 			to_chat(antag.current, "<img src='data:image/png;base64,[icon2base64(logo)]' style='position: relative; top: 10;'/> <span class='sinister'>[custom]</span>")
+		if ("converted")
+			to_chat(antag.current, "<span class='sinister'>You feel like you've broken past the veil of reality, your mind has seen worlds from beyond this plane, you've listened to the words of the Geometer of Blood for what felt like both an instant and ages, and now share both his knowledge and his ambition.</span>")
+			to_chat(antag.current, "<span class='sinister'>The Cult of Nar-Sie now counts you as its newest member. Your fellow cultists will guide you. You remember the last three words that Nar-Sie spoke to you: <span class='danger'>See Blood Hell</span></span>")
+		if ("pamphlet")
```
Comment (MadmanMartian) [inline code/datums/gamemode/role/cultist.dm]: DEFINES

**Review by MadmanMartian — CHANGES_REQUESTED** [review]

File: code/datums/gamemode/factions/bloodcult/bloodcult_items.dm (hunk 1 above)
Comment (DeityLink) [reply code/datums/gamemode/factions/bloodcult/bloodcult_items.dm]: hell if I know, this is part of the code I was moving around and I didn't check if all of it made sense.

File: code/datums/gamemode/factions/bloodcult/bloodcult_runespells.dm (hunk 10 above)
Comment (DeityLink) [reply code/datums/gamemode/factions/bloodcult/bloodcult_runespells.dm:676]: The value of "remaining" isn't increased, it's the progress done every second that is reduced. In other words, if you were to remove the implant through surgery mid conversion, then the rest of the conversion would progress much faster.

This check here is just so the flavor text is only sent once.

File: code/modules/mob/living/living.dm (hunk 9 above)
Comment (DeityLink) [reply code/modules/mob/living/living.dm]: >coding at 3am

File: code/datums/gamemode/factions/bloodcult/bloodcult_runespells.dm (hunk 11 above)
Comment (DeityLink) [reply code/datums/gamemode/factions/bloodcult/bloodcult_runespells.dm:694]: isn't that one set to 1 only after qdel is called? I might want this var to be set to 1 first.

File: code/datums/gamemode/factions/bloodcult/bloodcult_runespells.dm (hunk 10 above)
Comment (MadmanMartian) [reply code/datums/gamemode/factions/bloodcult/bloodcult_runespells.dm:676]: ooh, okay

File: code/datums/gamemode/factions/bloodcult/bloodcult_runespells.dm (hunk 11 above)
Comment (DamianX) [reply code/datums/gamemode/factions/bloodcult/bloodcult_runespells.dm:694]: I only saw that var being set in `Destroy`, so I assumed you just needed `gcDestroyed`. You can ignore this

File: code/datums/gamemode/factions/bloodcult/bloodcult.dm [hunk 14]
```diff
@@ -365,7 +365,26 @@ var/veil_thickness = CULT_PROLOGUE
...
 	data["blood"] = blood
 	return data
+
+
+/obj/item/proc/get_cult_power()
+	return 0
+
+/mob/proc/get_cult_power()
+	var/power = 0
+	var/list/valid_slots = list(
```
Comment (MadmanMartian) [inline code/datums/gamemode/factions/bloodcult/bloodcult.dm:379]: I've seen other collabs request for such lists that are just there as a pseudo-define to be defined as var/static/list so they don't need to keep being recreated each time a proc like this is called. Could you do so for this?

File: code/datums/gamemode/factions/bloodcult/bloodcult_runespells.dm [hunk 15]
```diff
@@ -537,23 +591,437 @@
...
+				if (victim.mind)
+					if (victim.mind.assigned_role in impede_medium)
+						progress = progress/2
+
+					if (victim.mind.assigned_role in impede_hard)
+						delay = 1
+						progress = progress/4
+
+				if (delay)
+					progress = max(1,min(10,progress))
```
Comment (MadmanMartian) [inline code/datums/gamemode/factions/bloodcult/bloodcult_runespells.dm:741]: Use Clamp instead?

File: code/modules/mob/living/living.dm [hunk 16]
```diff
@@ -1733,3 +1733,9 @@ Thanks.
 		"suiciding")
 
 	reset_vars_after_duration(resettable_vars, duration)
+
+/mob/living/proc/Silent(amount)
+	silent = max(max(silent,amount),0)
```
Comment (MadmanMartian) [inline code/modules/mob/living/living.dm:1738]: Why not have it be silent = max(SetSilent(amount),0) ?

Comment (DeityLink) [reply code/modules/mob/living/living.dm:1738]: SetSilent() doesn't return anything though

Comment (DeityLink) [reply code/modules/mob/living/living.dm:1738]: however I could replace it by:

SetSilent(max(silent,amount))

**Pull request merged**